code/modules/mob/living/carbon/give.dm

```diff
@@ -1,52 +1,167 @@
-		return
-	if((I.flags & NODROP) || (I.flags & ABSTRACT))
-		to_chat(usr, "<span class='notice'>That's not exactly something you can give.</span>")
-		return
-	if(target.r_hand == null || target.l_hand == null)
-		var/ans = alert(target,"[usr] wants to give you \a [I]?",,"Yes","No")
-		if(!I || !target)
-			return
-		switch(ans)
-			if("Yes")
-				if(target.incapacitated() || usr.incapacitated())
-					return
-				if(!Adjacent(target))
-					to_chat(usr, "<span class='warning'> You need to stay in reaching distance while giving an object.</span>")
-					to_chat(target, "<span class='warning'> [usr.name] moved too far away.</span>")
-					return
-				if((I.flags & NODROP) || (I.flags & ABSTRACT))
-					to_chat(usr, "<span class='warning'>[I] stays stuck to your hand when you try to give it!</span>")
-					to_chat(target, "<span class='warning'>[I] stays stuck to [usr.name]'s hand when you try to take it!</span>")
-					return
-				if(I != get_active_hand())
-					to_chat(usr, "<span class='warning'> You need to keep the item in your active hand.</span>")
-					to_chat(target, "<span class='warning'> [usr.name] seem to have given up on giving [I] to you.</span>")
-					return
-				if(target.r_hand != null && target.l_hand != null)
-					to_chat(target, "<span class='warning'> Your hands are full.</span>")
-					to_chat(usr, "<span class='warning'> Their hands are full.</span>")
-					return
-				usr.unEquip(I)
-				target.put_in_hands(I)
-				I.add_fingerprint(target)
-				target.visible_message("<span class='notice'> [usr.name] handed [I] to [target.name].</span>")
-				I.on_give(usr, target)
-			if("No")
-				target.visible_message("<span class='warning'> [usr.name] tried to hand [I] to [target.name] but [target.name] didn't want it.</span>")
-	else
-		to_chat(usr, "<span class='warning'> [target.name]'s hands are full.</span>")
+/**
+ * # Take Item alert
+ *
+ * Alert which appears for a user when another player is attempting to offer them an item.
+ * The user can click the alert to accept, or simply do nothing to not take the item.
+ */
+/obj/screen/alert/take_item
+	name = "Take Item"
+	desc = "someone wants to hand you an item!"
+	icon_state = "template"
+	timeout = 10 SECONDS
+	/// UID of the mob offering the receiver an item.
+	var/giver_UID
+	/// UID of the mob who has this alert.
+	var/receiver_UID
+	/// UID of the item being given.
+	var/item_UID
+
+
+/obj/screen/alert/take_item/Initialize(mapload, mob/giver, mob/receiver, obj/item/I)
+	. = ..()
+	desc = "[giver] wants to hand you \a [I]. Click here to accept it!"
+	giver_UID = giver.UID()
+	receiver_UID = receiver.UID()
+	item_UID = I.UID()
+	add_overlay(icon(I.icon, I.icon_state, SOUTH))
+	add_overlay("alert_flash")
+	ADD_TRAIT(giver, TRAIT_OFFERING_ITEM, TRAIT_GENERIC)
+	RegisterSignal(I, list(COMSIG_ITEM_EQUIPPED, COMSIG_ITEM_DROPPED), .proc/cancel_give)
+	RegisterSignal(giver, list(SIGNAL_ADDTRAIT(TRAIT_HANDS_BLOCKED), COMSIG_CARBON_SWAP_HANDS), .proc/cancel_give)
```

Comment:
I feel bad about doing nontrivial gameplay changes in `/obj/screen/alert/Initialize()`. It should be just an rectangle on the screen; giving it traits to people feels weird.

However, i don't have a good idea how to fix this outside of a whole new `/datum/gifting_process` that would manage this, which feels terrifically overengineered (and probably worse of the two evils).
Sad.

Reply:
Yep, I thought about using a datum briefly, but it just makes things more annoying.